src/Native/Unix/System.Native/pal_random.c

```diff
@@ -2,109 +2,258 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-#include <stdlib.h>
-#include <stdint.h>
+#include <assert.h>
+#include <errno.h>
+#include <fcntl.h>
 #include <stdbool.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
 #include <sys/stat.h>
-#include <fcntl.h>
-#include <assert.h>
-#include <unistd.h>
 #include <time.h>
-#include <errno.h>
+#include <unistd.h>
 
 #include "pal_config.h"
 #include "pal_random.h"
+
+#if defined(__linux__)
+#include <sys/syscall.h>
+#if defined(SYS_getrandom)
+# define PAL_GRND_NONBLOCK 1
+#endif // SYS_getrandom
+#endif // __linux__
+
 /*
 
 Generate random bytes. The generated bytes are not cryptographically strong.
 
 */
 
-void SystemNative_GetNonCryptographicallySecureRandomBytes(uint8_t* buffer, int32_t bufferLength)
+#if !HAVE_ARC4RANDOM_BUF
+static int GetUrandomFileDescriptor(void)
 {
-    assert(buffer != NULL);
-
-#if HAVE_ARC4RANDOM_BUF
-    arc4random_buf(buffer, (size_t)bufferLength);
-#else
-    static volatile int rand_des = -1;
-    long num = 0;
-    static bool sMissingDevURandom;
-    static bool sInitializedMRand;
+    static int fd = -1;
+    int newFd;
 
-    if (!sMissingDevURandom)
+    if (fd >= 0)
     {
-        if (rand_des == -1)
-        {
-            int fd;
+        return fd;
+    }
 
-            do
-            {
+    do
+    {
 #if HAVE_O_CLOEXEC
-                fd = open("/dev/urandom", O_RDONLY, O_CLOEXEC);
+        newFd = open("/dev/urandom", O_RDONLY | O_CLOEXEC);
 #else
-                fd = open("/dev/urandom", O_RDONLY);
-                fcntl(fd, F_SETFD, FD_CLOEXEC);
+        newFd = open("/dev/urandom", O_RDONLY);
+        fcntl(newFd, F_SETFD, FD_CLOEXEC);
 #endif
-            }
-            while ((fd == -1) && (errno == EINTR));
+    }
+    while ((newFd == -1) && (errno == EINTR));
+
+    int unsetFd = -1;
+    if (!__atomic_compare_exchange_n(&fd, &unsetFd, newFd, false, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST))
+    {
+        close(newFd);
+    }
+
+    return fd;
+}
+
+static void XorWithLrand48(uint8_t *buffer, int32_t bufferLength)
+{
+    static bool sInitializedLrand;
+    size_t length = (size_t)bufferLength;
+
+    if (!sInitializedLrand)
+    {
+        bool lRandUninitialized = false;
+
+        if (__atomic_compare_exchange_n(&sInitializedLrand, &lRandUninitialized, true, false,
```

Comment:
This sets the flag first and then does the initialization. Shouldn't it be the other way around?

Reply:
I think that it's trying to avoid the side effect of calling srand a second time.

Seems like it does need some sort of wait for the race losers so they don't call lrand48 before it's seeded.

```C

if (!sInitializedLrand)
{
    static bool sShouldIDoIt;
    bool lShouldIDoIt = false;

    if (atomic exchange(shouldIDoIt))
    {
        srand48(time(NULL));
        sInitializedLrand = true;
    }
    else
    {
        while (!sInitializedLrand)
        {
            sleep(1);
        }
    }
}
```

?

Reply:
Indeed, nice catch.  Will re-push a fix shortly after this builds and I'm able to test by intentionally introducing a race (adding a sleep() in the case where the CAS succeeds).